### PR TITLE
Take BUCKET_ID column into account when applying delete_delta

### PR DESCRIPTION
**Pull request by losipiuk**
Fix a bug when it was possible that more rows than expected
where excluded from a query result, when delete_delta contained two
rows with same ROW_ID/ORIGINAL_TRANSACTION pair but a different BUCKET_ID.
It was possible a single row was deleted from multiple rows created
with a Hive multi-insert statement.

See https://issues.apache.org/jira/browse/HIVE-16832 for more context

Bug was introduced as part of 62e1622, yet the fix does not exactly
restore previous logic which is no longer strictly applicable.

### CODE REVIEW

**Comment by losipiuk** [comment]
Code credits @alexjo2144

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 1]
```diff
@@ -183,17 +191,28 @@ private RowId getRowId(int position)
...
             if (startRowId.isPresent()) {
                 // original transaction ID is always 0 for original file row delete delta.
                 originalTransaction = 0;
                 // In case of original files, calculate row ID is start row ID of the page + current position in the page
                 row = startRowId.getAsLong() + position;
+                // For original files set the bucket number to original bucket number if table was bucketed or to 0 if it was not.
+                // Set statement Id to 0.
+                // Verified manually that this is consistent with Hive 3.1 behavior.
+                bucket = bucketNumber.orElse(0);
+                statementId = 0;
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:205]: move before `row = ...` line (maintain ordering "original transaction, bucket, statement id, row id")

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 2]
```diff
@@ -183,17 +191,28 @@ private RowId getRowId(int position)
...
             }
             else {
                 originalTransaction = BIGINT.getLong(sourcePage.getBlock(ORIGINAL_TRANSACTION_INDEX), position);
                 row = BIGINT.getLong(sourcePage.getBlock(ROW_ID_INDEX), position);
+                int bucketData = toIntExact(INTEGER.getLong(sourcePage.getBlock(BUCKET_ID_INDEX), position));
+                BucketCodec bucketCodec = BucketCodec.determineVersion(bucketData);
+                bucket = bucketCodec.decodeWriterId(bucketData);
+                statementId = bucketCodec.decodeStatementId(bucketData);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:213]: move before `row = ...` line

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java [hunk 3]
```diff
@@ -217,8 +236,13 @@ private RowId getRowId(int position)
                         if (page != null) {
                             for (int i = 0; i < page.getPositionCount(); i++) {
                                 long originalTransaction = BIGINT.getLong(page.getBlock(ORIGINAL_TRANSACTION_INDEX), i);
+                                int bucketData = toIntExact(INTEGER.getLong(page.getBlock(BUCKET_ID_INDEX), i));
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:239]: "bucketValue"?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 4]
```diff
@@ -612,6 +612,33 @@ public void testMultiDelete(HiveOrPresto inserter, HiveOrPresto deleter)
         });
     }
 
+    @Test(groups = HIVE_TRANSACTIONAL, timeOut = TEST_TIMEOUT)
+    public void testMultiInsertWithDelete()
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:616]: `testReadAfterMultiInsertAndDelete` ?

Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:616]: Maybe add a javadoc sth like

```
Test reading from a table after Hive multi-insert. Multi-insert involves non-zero statement ID, encoded
within ORC-level bucketId field, while originalTransactionId and rowId can be the same for two different rows.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 5]
```diff
@@ -612,6 +612,33 @@ public void testMultiDelete(HiveOrPresto inserter, HiveOrPresto deleter)
...
+        withTemporaryTable("partitioned_multi_insert", true, true, BucketingType.BUCKETED_V1, tableName -> {
+            withTemporaryTable("tmp_data_table", false, false, NONE, dataTableName -> {
+                onTrino().executeQuery(format("CREATE TABLE %s (a int, b int, c varchar(5)) WITH " +
+                        "(transactional = true, partitioned_by = ARRAY['c'], bucketed_by = ARRAY['a'], bucket_count = 2)", tableName));
+                onTrino().executeQuery(format("CREATE TABLE %s (x int)", dataTableName));
+                onTrino().executeQuery(format("INSERT INTO %s VALUES 1", dataTableName));
+
+                // Perform a multi-insert
+                onHive().executeQuery("SET hive.exec.dynamic.partition.mode = nonstrict");
+                onHive().executeQuery(format("FROM %s INSERT INTO %s partition(c) SELECT 0, 0, 'c' || x INSERT INTO %2$s partition(c='c1') SELECT 0, 1",
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:629]: Maybe add a comment why mixing dynamic partition with explicit partition?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 6]
```diff
@@ -612,6 +612,33 @@ public void testMultiDelete(HiveOrPresto inserter, HiveOrPresto deleter)
...
+                        dataTableName,
+                        tableName));
+                onHive().executeQuery(format("DELETE FROM %s WHERE b = 1", tableName));
+
+                List<QueryAssert.Row> expected = ImmutableList.of(row(0, 0, "c1"));
+                String selectQuery = "SELECT * FROM " + tableName;
+                assertThat(query(selectQuery)).containsExactly(expected);
+                assertThat(onHive().executeQuery(selectQuery)).containsExactly(expected);
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:637]: use `verifySelectForPrestoAndHive` and inline the `expected` and `selectQuery`

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java (hunk 3 above)
Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcDeletedRows.java:239]: Renamed to `encodedBucketValue` here an above

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 5 above)
Comment (losipiuk) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:629]: I added
```
// Combine dynamic and static partitioning to trick Hive to insert two rows with same rowId but different statementId to a single partition.
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 4 above)
Comment (losipiuk) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:616]: Added in method body. And restructured the comment we already had a bit.

**Comment by losipiuk** [comment]
I needed to add https://github.com/trinodb/trino/pull/8208/commits/3a3c2dc62756073e9de6a3d02a5863d9af81e278.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java [hunk 7]
```diff
@@ -562,8 +562,8 @@ static void verifyAcidSchema(OrcReader orcReader, Path path)
                             path));
         }
         verifyAcidColumn(orcReader, 0, AcidSchema.ACID_COLUMN_OPERATION, INT, path);
-        verifyAcidColumn(orcReader, 1, AcidSchema.ACID_COLUMN_ORIGINAL_TRANSACTION, LONG, path);
         verifyAcidColumn(orcReader, 2, AcidSchema.ACID_COLUMN_BUCKET, INT, path);
+        verifyAcidColumn(orcReader, 1, AcidSchema.ACID_COLUMN_ORIGINAL_TRANSACTION, LONG, path);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:567]: please keep lines in 0, 1, 2, .. order

Comment (losipiuk) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:567]: Change in this line did not make any sense .... reverted

**Review by findepi — APPROVED** [review]

**Pull request merged**